spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/BigQueryRDDFactory.java

```diff
@@ -0,0 +1,184 @@
+ * limitations under the License.
+ */
+
+package com.google.cloud.spark.bigquery;
+
+import com.google.cloud.bigquery.Field;
+import com.google.cloud.bigquery.Schema;
+import com.google.cloud.bigquery.StandardTableDefinition;
+import com.google.cloud.bigquery.TableDefinition;
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.bigquery.TableInfo;
+import com.google.cloud.bigquery.connector.common.BigQueryClient;
+import com.google.cloud.bigquery.connector.common.BigQueryClientFactory;
+import com.google.cloud.bigquery.connector.common.BigQueryUtil;
+import com.google.cloud.bigquery.connector.common.ReadSessionCreator;
+import com.google.cloud.bigquery.connector.common.ReadSessionResponse;
+import com.google.cloud.bigquery.storage.v1.ReadSession;
+import com.google.cloud.spark.bigquery.direct.BigQueryRDD;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Streams;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.spark.rdd.RDD;
+import org.apache.spark.sql.SQLContext;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Wrapper class for generating BigQueryRDD. Extracted this logic out so that we can reuse it from
+ * 1) Dsv1 buildScan 2) Dsv1 pushdown functionality 3) Dsv2 pushdown functionality
+ */
+public class BigQueryRDDFactory {
+  /**
+   * Default parallelism to 1 reader per 400MB, which should be about the maximum allowed by the
+   * BigQuery Storage API. The number of partitions returned may be significantly less depending on
+   * a number of factors.
+   */
+  private static long DEFAULT_BYTES_PER_PARTITION = 400L * 1000 * 1000;
+
+  private static final Logger log = LoggerFactory.getLogger(BigQueryRDDFactory.class);
+
+  private final BigQueryClient bigQueryClient;
+  private final SparkBigQueryConfig options;
+  private final BigQueryClientFactory bigQueryReadClientFactory;
+  private final SQLContext sqlContext;
+
+  public BigQueryRDDFactory(
+      BigQueryClient bigQueryClient,
+      BigQueryClientFactory bigQueryReadClientFactory,
+      SparkBigQueryConfig options,
+      SQLContext sqlContext) {
+    this.bigQueryClient = bigQueryClient;
+    this.options = options;
+    this.bigQueryReadClientFactory = bigQueryReadClientFactory;
+    this.sqlContext = sqlContext;
+  }
+
+  /**
+   * Creates RDD from the SQL string that is passed in. This functionality is invoked from the query
+   * pushdown module
+   */
+  public RDD<InternalRow> buildScanFromSQL(String sql) {
+    log.info("Materializing the following sql query to a BigQuery table: {}", sql);
+
+    TableInfo actualTable =
+        bigQueryClient.materializeQueryToTable(
+            sql, options.getMaterializationExpirationTimeInMinutes());
+    TableDefinition actualTableDefinition = actualTable.getDefinition();
+
+    List<String> requiredColumns =
```

Comment:
Not sure this is needed. Passing an empty list on CreateReadSession is equivalent to `SELECT *`.

Reply:
Oh interesting. Good call